crates/rules/src/roles/builder/roles_builder.rs

```diff
@@ -0,0 +1,603 @@
+use crate::prelude::*;
+
+impl RoleBuilder {
+    pub fn primary() -> Self {
+        Self::new(RoleKind::Primary)
+    }
+
+    pub fn recovery() -> Self {
+        Self::new(RoleKind::Recovery)
+    }
+
+    pub fn confirmation() -> Self {
+        Self::new(RoleKind::Confirmation)
+    }
+}
+
+#[cfg(test)]
+impl RoleWithFactorSourceIds {
+    pub(crate) fn primary_with_factors(
+        threshold: u8,
+        threshold_factors: impl IntoIterator<Item = FactorSourceID>,
+        override_factors: impl IntoIterator<Item = FactorSourceID>,
+    ) -> Self {
+        Self::with_factors(
+            RoleKind::Primary,
+            threshold,
+            threshold_factors,
+            override_factors,
+        )
+    }
+
+    pub(crate) fn recovery_with_factors(
+        override_factors: impl IntoIterator<Item = FactorSourceID>,
+    ) -> Self {
+        Self::with_factors(RoleKind::Recovery, 0, vec![], override_factors)
+    }
+
+    pub(crate) fn confirmation_with_factors(
+        override_factors: impl IntoIterator<Item = FactorSourceID>,
+    ) -> Self {
+        Self::with_factors(RoleKind::Confirmation, 0, vec![], override_factors)
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, thiserror::Error)]
+pub enum RoleBuilderValidation {
+    #[error("Basic violation: {0}")]
+    BasicViolation(#[from] BasicViolation),
+
+    #[error("Forever invalid: {0}")]
+    ForeverInvalid(#[from] ForeverInvalidReason),
+
+    #[error("Not yet valid: {0}")]
+    NotYetValid(#[from] NotYetValidReason),
+}
+type Validation = RoleBuilderValidation;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, thiserror::Error)]
+pub enum BasicViolation {
+    /// e.g. tried to remove a factor source which was not found.
+    #[error("FactorSourceID not found")]
+    FactorSourceNotFound,
+
+    #[error("Recovery cannot set threshold")]
+    RecoveryCannotSetThreshold,
+
+    #[error("Confirmation cannot set threshold")]
+    ConfirmationCannotSetThreshold,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, thiserror::Error)]
+pub enum NotYetValidReason {
```

Comment:
In earlier implementations I had one `error` enum type for each Role, maybe that makes sense, not sure. It results in many more types and more complex translation into `RoleBuilderValidation` so I deemed it not worth it. 

Furtermore, all these violations are "ephemeral" in the sense that they ALL will be translated into `CommonError` at some point in time...